### PR TITLE
Immutable arrays

### PR DESCRIPTION
**Pull request by goldfirere**
I've created this PR instead of working in https://github.com/ocaml-flambda/ocaml-jst/pull/47 so that the comprehensions stuff doesn't get in the way.

### CODE REVIEW

File: stdlib/array.ml [hunk 1]
```diff
@@ -162,6 +162,15 @@ let of_list = function
         | hd::tl -> unsafe_set a i hd; fill (i+1) tl in
       fill 1 tl
 
+(* Immutable and mutable arrays have the same runtime representation; we
```
Comment (goldfirere) [inline stdlib/array.ml:165]: @mshinwell, could you confirm that this is OK? I see that mutability is tracked in Lambda, but indeed it seems not to change the translation away from Lambda, so this is likely OK. But `Obj.magic` requires extra scrutiny. Thanks!

File: stdlib/array.mli [hunk 2]
```diff
@@ -24,6 +24,13 @@ open! Stdlib
    arrayLabels.mli instead.
  *)
 
+(* CR aspectorzabusky: When I ran [tools/sync_stdlib_docs], it copied the
```
Comment (goldfirere) [inline stdlib/array.mli:27]: @stedolan, I'm sure you know the answer to this...

File: stdlib/array.ml [hunk 3]
```diff
@@ -162,6 +162,15 @@ let of_list = function
...
+   construct immutable arrays by constructing mutable arrays and then blindly
+   casting them to become immutable.  This is safe here because [copy]:
+   1. doesn't mutate its input;
+   2. doesn't hold on to its input; and
+   3. returns a fresh array as its output. *)
+let to_iarray = Obj.magic (copy : 'a array -> 'a array)
```
Comment (goldfirere) [inline stdlib/array.ml:171]: Maybe I'm just paranoid, but I'd prefer a `let to_iarray : 'a array -> 'a iarray = ...` here. Yes, it's redundant with the mli file. But it would be Bad to change the mli file's type signature and then quietly make this unsound.

File: stdlib/array.ml [hunk 4]
```diff
@@ -162,6 +162,15 @@ let of_list = function
...
+let of_iarray = Obj.magic (copy : 'a array -> 'a array)
```
Comment (goldfirere) [inline stdlib/array.ml:172]: These functions make me want

```ocaml
type immutable = |
type mutable = |
type ('a, 'mut) gen_array

type 'a array = ('a, mutable) gen_array
type 'a iarray = ('a, immutable) gen_array
```

and then many functions (including user-written ones) can be polymorphic over mutability. This also means that we could drop the `Obj.magic` here.

Actually, this makes me think: immutability could be done entirely in a library. The only reason to involve the compiler at all is 1) to perform optimizations and 2) to provide good syntax. Would this approach simplify things?

File: stdlib/iarray.ml [hunk 5]
```diff
@@ -0,0 +1,51 @@
+(* CR aspectorzabusky: This needs a copyright header; should I copy [array.ml]? *)
```
Comment (goldfirere) [inline stdlib/iarray.ml:1]: @lpw25 any thoughts?

File: stdlib/iarray.ml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+open! Stdlib
+
+[@@@ocaml.flambda_o3]
+
+(* An alias for the type of arrays. *)
```
Comment (goldfirere) [inline stdlib/iarray.ml]:
```suggestion
(* An alias for the type of immutable arrays. *)
```

File: stdlib/iarrayLabels.mli [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+(** [init n ~f] returns a fresh immutable array of length [n],
+   with element number [i] initialized to the result of [f i].
+   In other terms, [init n ~f] tabulates the results of [f]
+   applied to the integers [0] to [n-1].
+
+   @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
+   If the return type of [f] is [float], then the maximum
+   size is only [Sys.max_array_length / 2]. *)
+
+val make_matrix :
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli]: This function seems pretty unuseful.

File: stdlib/iarrayLabels.mli [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+val sub : 'a iarray -> pos:int -> len:int -> 'a iarray
+(** [sub a ~pos ~len] returns a fresh immutable array of length [len],
+   containing the elements number [pos] to [pos + len - 1]
+   of immutable array [a].
+
+   @raise Invalid_argument if [pos] and [len] do not
+   designate a valid subarray of [a]; that is, if
+   [pos < 0], or [len < 0], or [pos + len > length a]. *)
+
+(* CR aspectorzabusky: I dropped [copy] because these are immutable.  Is there
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli]: No, I don't think so. Leave it out.

File: stdlib/iarrayLabels.mli [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+val append : 'a iarray -> 'a iarray -> 'a iarray
+(** [append v1 v2] returns a fresh immutable array containing the
+   concatenation of the immutable arrays [v1] and [v2].
+   @raise Invalid_argument if
+   [length v1 + length v2 > Sys.max_array_length]. *)
+
+val concat : 'a iarray list -> 'a iarray
+(** Same as {!append}, but concatenates a list of immutable arrays. *)
+
+val sub : 'a iarray -> pos:int -> len:int -> 'a iarray
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli:60]: Does this copy? It does not need to. Actually, it's a bit of a shame that it does copy. The ability to slice an immutable array is one of its advantages over a mutable one. In any case, if it copies, the documentation should say so.

File: stdlib/iarrayLabels.mli [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+   @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
+   If the value of [x] is a floating-point number, then the maximum
+   size is only [Sys.max_array_length / 2]. *)
+
+external create_float: int -> float iarray = "caml_make_float_vect"
+(** [create_float n] returns a fresh immutable float array of length [n],
+    with uninitialized data. *)
+
+val init : int -> f:(int -> 'a) -> 'a iarray
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli:41]: The `f` argument can be marked as local. Actually, a lot of arguments here can be local. Perhaps it would be good to imagine how to make this interface more local-friendly. Could also be a good topic for a type-systems meeting.

File: stdlib/iarrayLabels.mli [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+   If this file is iarrayLabels.mli, run tools/sync_stdlib_docs after editing it
+   to generate iarray.mli.
+
+   If this file is iarray.mli, do not edit it directly -- edit
+   iarrayLabels.mli instead.
+ *)
+
+(** Operations on immutable arrays. *)
+
+type +'a t = 'a iarray
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli:23]: Is it possible to put an array on the stack? If so, it would be good to annotate locality information on the functions here.

File: stdlib/iarrayLabels.mli [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+val of_list : 'a list -> 'a iarray
+(** [of_list l] returns a fresh immutable array containing the elements
+   of [l].
+
+   @raise Invalid_argument if the length of [l] is greater than
+   [Sys.max_array_length]. *)
+
+(** {1 Converting to and from mutable arrays} *)
+
+(* CR aspectorzabusky: When we add locals, we can do
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli]: We have added locals!

File: stdlib/iarrayLabels.mli [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+   The last element has number [length a - 1].
+   You can also write [a.:(n)] instead of [get a n].
+
+   @raise Invalid_argument
+   if [n] is outside the range 0 to [(length a - 1)]. *)
+
+external ( .:() ) : 'a iarray -> int -> 'a = "%array_safe_get"
+(** A synonym for [get]. *)
+
+external make : int -> 'a -> 'a iarray = "caml_make_vect"
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli]: This seems not very useful.

File: stdlib/iarrayLabels.mli [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+   physically equal to [x] (in the sense of the [==] predicate).
+   Consequently, if [x] is mutable, it is shared among all elements
+   of the array, and modifying [x] through one of the array entries
+   will modify all other entries at the same time.
+
+   @raise Invalid_argument if [n < 0] or [n > Sys.max_array_length].
+   If the value of [x] is a floating-point number, then the maximum
+   size is only [Sys.max_array_length / 2]. *)
+
+external create_float: int -> float iarray = "caml_make_float_vect"
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli]: This seems positively unuseful.

File: stdlib/iarrayLabels.mli [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+(** [map ~f a] applies function [f] to all the elements of [a],
+   and builds an immutable array with the results returned by [f]:
+   [[| f a.:(0); f a.:(1); ...; f a.:(length a - 1) |]]. *)
+
+val mapi : f:(int -> 'a -> 'b) -> 'a iarray -> 'b iarray
+(** Same as {!map}, but the
+   function is applied to the index of the element as first argument,
+   and the element itself as second argument. *)
+
+val fold_left : f:('a -> 'b -> 'a) -> init:'a -> 'b iarray -> 'a
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli:112]: 4.14 also comes with `fold_left_map`, which should be added here, too.

File: stdlib/iarrayLabels.mli [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+
+val mem : 'a -> set:'a iarray -> bool
+(** [mem a ~set] is true if and only if [a] is structurally equal
+    to an element of [l] (i.e. there is an [x] in [l] such that
+    [compare a x = 0]). *)
+
+val memq : 'a -> set:'a iarray -> bool
+(** Same as {!mem}, but uses physical equality
+   instead of structural equality to compare list elements. *)
+
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli:174]: Also add 4.14's `find_opt`, `find_map`, `split`, and `combine`.

File: stdlib/iarrayLabels.mli [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+(* CR aspectorzabusky: We should add non–in-place sorting *)
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli]: Yes.

File: stdlib/iarrayLabels.mli [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+
```
Comment (goldfirere) [inline stdlib/iarrayLabels.mli:236]: Might be nice to have some kind of

```ocaml
(** Create an immutable array by filling in a mutable array. This is the same as building a mutable array
and then using [of_array], but without a copy. *)
val make_mutably : int -> (local_ 'a array -> unit) -> 'a iarray
```

This would then allow users to e.g. write `sort`. Note that without the `local_`, this would be wildly unsafe! But I think it's safe with the `local_`, given that all mutable fields must be global. (The closure itself could also be labeled as `local_`, but that's a separate matter.)

File: typing/outcometree.mli [hunk 19]
```diff
@@ -40,6 +40,7 @@ type out_attribute =
 
 type out_value =
   | Oval_array of out_value list
+  | Oval_iarray of out_value list (* CR aspectorzabusky: Or [Oval_immutable_array]? *)
```
Comment (goldfirere) [inline typing/outcometree.mli]: `iarray` is fine. But actually, I think it would be slightly better to have `Oval_array of out_value list * Asttypes.mutable_flag`.

File: typing/parmatch.ml [hunk 20]
```diff
@@ -302,7 +302,9 @@ module Compat
   | Tpat_record (l1,_),Tpat_record (l2,_) ->
       let ps,qs = records_args l1 l2 in
       compats ps qs
-  | Tpat_array ps, Tpat_array qs ->
+  | Tpat_array (am1, ps), Tpat_array (am2, qs) ->
+      am1 = am2 &&
+      (* CR aspectorzabusky: Can't we drop the length check here? *)
```
Comment (goldfirere) [inline typing/parmatch.ml]: I don't think so. Two patterns are compatible if there exists a value that matches both, as per "Compatibility check" above.

File: typing/typecore.ml [hunk 21]
```diff
@@ -1683,6 +1691,32 @@ and type_pat_aux
...
+      pat_mode = alloc_mode.mode;
+      pat_attributes = sp.ppat_attributes;
+      pat_env = !env })
+  in
+  match Extensions.Pattern.of_ast sp with
+  | Some epat -> begin
+      (* Normally this would go to an auxiliary function, but this function
+         takes so many parameters, has such a complex type, and uses so many
+         local definitions, it seems better to just put the pattern matching
+         here.  This shouldn't fuck up the diff *too* much. *)
```
Comment (goldfirere) [inline typing/typecore.ml]: Maybe I'm just a bit curmudgeonly, but I'd prefer cleaner language in the codebase.

File: typing/typecore.ml [hunk 22]
```diff
@@ -2890,6 +2918,13 @@ let is_local_returning_expr e =
         raise(Error(loc2, Env.empty, Local_return_annotation_mismatch loc1))
   in
   let rec loop e =
+    match Extensions.Expression.of_ast e with
```
Comment (goldfirere) [inline typing/typecore.ml:3163]: Seeing this inspired me to look for other `Parsetree.expression` matches. I found one in `parsing/ast_invariants.ml` (line 80) that should probably get a call to `of_ast`.

File: typing/typecore.ml [hunk 23]
```diff
@@ -5896,7 +5931,7 @@ and type_let
       ({ pexp_desc = Pexp_extension({txt = "extension.local"}, PStr []) },
        [Nolabel, e]) -> sexp_is_fun e
     | _ -> false
-  and eexp_is_fun : Extensions.extension_expr -> _ = function
+  and eexp_is_fun : Extensions.Expression.t -> _ = function
     | _ -> false
```
Comment (goldfirere) [inline typing/typecore.ml]: Maybe avoid the `_ ->` here?

File: typing/typecore.ml [hunk 24]
```diff
@@ -6510,6 +6585,20 @@ and type_comprehension_iterator
...
+  | Iaexp_immutable_array elts ->
+      type_generic_array
+        ~loc
+        ~env
+        ~expected_mode
+        ~ty_expected
+        ~explanation
+        ~type_:Predef.type_iarray
+        ~mutability:Immutable
+        ~attributes:[] (* CR aspectorzabusky: This can't be right *)
```
Comment (goldfirere) [inline typing/typecore.ml:7098]: You're right -- this is problematic. The problem is in the parsing scheme, which doesn't really put the attributes anywhere.

File: typing/typecore.ml [hunk 25]
```diff
@@ -6656,8 +6745,11 @@ let report_type_expected_explanation expl ppf =
   | Comprehension_in_iterator comp_ty ->
       because1 "a for-in iterator in %s comprehension"
         (match comp_ty with
-         | List_comprehension  -> "a list"
-         | Array_comprehension -> "an array")
+         | List_comprehension   -> "a list"
+         | Array_comprehension  -> "an array" (* CR aspectorzabusky: We should
```
Comment (goldfirere) [inline typing/typecore.ml]: Yes, skip "mutable".

File: typing/typecore.mli [hunk 26]
```diff
@@ -23,6 +23,7 @@ open Types
 type comprehension_type =
   | List_comprehension
   | Array_comprehension
```
Comment (goldfirere) [inline typing/typecore.mli]: It's not clear to me whether this is better or `Array_comprehension of Asttypes.mutable_flag` is better. Just throwing out another possibility -- it's perfectly fine to keep this as-is, too.

File: typing/typedtree.mli [hunk 27]
```diff
@@ -113,8 +113,13 @@ and 'k pattern_desc =
 
             Invariant: n > 0
          *)
-  | Tpat_array : value general_pattern list -> value pattern_desc
-        (** [| P1; ...; Pn |] *)
+  | Tpat_array :
+      mutable_flag * value general_pattern list -> value pattern_desc
+        (** [| P1; ...; Pn |]    (flag = Mutable)
+            [: P1; ...; Pn :]    (flag = Immutable) *)
+    (* CR aspectorzabusky: This ended up feeling clearer to me than two
```
Comment (goldfirere) [inline typing/typedtree.mli]: I like this design

File: typing/typedtree.mli [hunk 28]
```diff
@@ -248,9 +253,13 @@ and expression_desc =
   | Texp_field of expression * Longident.t loc * Types.label_description
   | Texp_setfield of
       expression * Longident.t loc * Types.label_description * expression
-  | Texp_array of expression list
+  | Texp_array of mutable_flag * expression list
+    (* CR aspectorzabusky: This ended up feeling clearer to me than two
```
Comment (goldfirere) [inline typing/typedtree.mli]: I like this design. If anything, I would want more of this kind of design for this feature, though you have the experience to make the better decision.

File: lambda/transl_array_comprehension.mli [hunk 29]
```diff
@@ -4,12 +4,17 @@ open Debuginfo.Scoped_location
 
 (** Translate array comprehensions; see the .ml file for more details *)
 
+(* CR aspectorzabusky: Is it really true that I don't need to pass in array
```
Comment (goldfirere) [inline lambda/transl_array_comprehension.mli:7]: @mshinwell? This code, as written, will result in `Pmakearray(..., Mutable, ...)`, even for an immutable array. This makes sense, because the comprehensions code actually mutates the memory after creation. But then we'll have a mutable block as the value for an immutable array. I don't quite see what goes wrong here, but I'd love more scrutiny on this point. (@antalsz: Given that Mark is away for the next few weeks, feel free to find someone else who can answer this question.)

File: lambda/translcore.ml [hunk 30]
```diff
@@ -31,7 +31,7 @@ type error =
 
 exception Error of Location.t * error
 
-let use_dup_for_constant_arrays_bigger_than = 4
+let use_dup_for_constant_mutable_arrays_bigger_than = 4
```
Comment (goldfirere) [inline lambda/translcore.ml:35]: I'm scared of this module, because
1. I don't know the ramifications of the `Mutable`/`Immutable` flag in lambda.
2. Getting this wrong seems like it could be catastrophic.
@mshinwell is an obvious reviewer for this module, but others may also do well.

File: parsing/extensions.ml [hunk 31]
```diff
@@ -1,265 +1,46 @@
...
-    these forms directly.  They're just an implementation detail.
-
-    Within this module, we provide some simple machinery for working with these
-    [([%extension.EXTNAME.ID1.ID2.….IDn] EXPR)] wrapped forms.  To construct
-    one, we provide [extension_expr]; to destructure one, we provide
-    [expand_extension_expr].  We still have to write the transformations in both
-    directions for all new syntax, lowering it to extension nodes and then
-    (somewhat more obnoxiously) lifting it back out. *)
-
+open Asttypes
```
Comment (goldfirere) [inline parsing/extensions.ml:1]: I am not reviewing the extensions architecture. See #1.

File: parsing/parser.mly [hunk 32]
```diff
@@ -2351,6 +2378,7 @@ simple_expr:
       { array_get ~loc:$sloc $1 $4 }
   | simple_expr DOT LPAREN seq_expr error
       { unclosed "(" $loc($3) ")" $loc($5) }
+  (* Immutable array indexing is a regular operator *)
```
Comment (goldfirere) [inline parsing/parser.mly]: I don't understand the import of this comment.

File: parsing/parser.mly [hunk 33]
```diff
@@ -2496,20 +2542,20 @@ comprehension_clause:
...
-  | od=open_dot_declaration DOT LBRACKETBAR expr_semi_list BARRBRACKET
-      { Pexp_open(od, mkexp ~loc:($startpos($3), $endpos) (Pexp_array($4))) }
-  | od=open_dot_declaration DOT LBRACKETBAR BARRBRACKET
-      { (* TODO: review the location of Pexp_array *)
-        Pexp_open(od, mkexp ~loc:($startpos($3), $endpos) (Pexp_array [])) }
-  | mod_longident DOT
-    LBRACKETBAR expr_semi_list error
-      { unclosed "[|" $loc($3) "|]" $loc($5) }
+  | array_exprs(LBRACKETBAR, BARRBRACKET)
+      { Generic_array.expression
```
Comment (goldfirere) [inline parsing/parser.mly:2665]: Is `Generic_array` really necessary here? Can you not pass ml values into a parameterized non-terminal? What if you had a function that branched on the open/close terminal? If `Generic_array` stays, though, then it needs more documentation.

File: stdlib/array.ml (hunk 4 above)
Comment (goldfirere) [reply stdlib/array.ml:172]: I suppose this would destroy the nice covariance annotation. But is that worth it? Or maybe I'm missing something more fundamental.

**Review by goldfirere — COMMENTED** [review]
Looks pretty good! There are a few loose ends, as you commented, and a few parts I was not competent to review.

File: stdlib/array.ml (hunk 4 above)
Comment (goldfirere) [reply stdlib/array.ml:172]: After more thought, this is a move in the wrong direction: covariance is nice indeed, and there is the possibility for efficient array slicing for immutable arrays.

File: typing/parmatch.ml (hunk 20 above)
Comment (antalsz) [reply typing/parmatch.ml]: It would be correct to delete the length check but it's an optimization, because arrays have O(1) length

File: stdlib/iarrayLabels.mli (hunk 7 above)
Comment (antalsz) [reply stdlib/iarrayLabels.mli]: Done, and I added a comment about it.

File: stdlib/array.ml (hunk 3 above)
Comment (antalsz) [reply stdlib/array.ml:171]: This would have been nice, but ran into the value restriction.  I've added a comment instead.

File: typing/typecore.ml (hunk 22 above)
Comment (antalsz) [reply typing/typecore.ml:3163]: This actually raises an interesting can of worms: what do we do with `ast_iterator`?  I think we have to leave it alone, right?

File: typing/typecore.mli (hunk 26 above)
Comment (antalsz) [reply typing/typecore.mli]: I think moving pretty uniformly to the `mutable_flag` form is better; change made

File: typing/typedtree.mli (hunk 27 above)
Comment (antalsz) [reply typing/typedtree.mli]: Yes, moving this way everywhere.

File: typing/typedtree.mli (hunk 28 above)
Comment (antalsz) [reply typing/typedtree.mli]: I think you're correct here.

File: parsing/parser.mly (hunk 32 above)
Comment (antalsz) [reply parsing/parser.mly]: This should've gotten clarified in the rebase, there are a couple new comments about immutable array indexing now

File: typing/typecore.ml (hunk 22 above)
Comment (goldfirere) [reply typing/typecore.ml:3163]: "Have to"? No. We could add a little hook that processes immutable arrays in some way. But there is no good reason to do so -- note that arrays simply process their subexpressions. So we should do the same for immutable arrays, which happens automatically.

Comment (antalsz) [reply typing/typecore.ml:3163]: Also, fixing this triggered a test failure; please take a look at the solution in `testsuite/tests/ast-invariants/test.ml`, I'm not sure if it's the best approach

File: typing/typecore.ml (hunk 24 above)
Comment (antalsz) [reply typing/typecore.ml:7098]: This will be dealt with in the extensions PR

File: stdlib/iarrayLabels.mli (hunk 9 above)
Comment (antalsz) [reply stdlib/iarrayLabels.mli:60]: As discussed, it does have to copy; I've added a sentence saying so.

File: stdlib/iarrayLabels.mli (hunk 10 above)
Comment (antalsz) [reply stdlib/iarrayLabels.mli:41]: We decided not to make these functions local in the stdlib and only do that in Base

File: stdlib/iarrayLabels.mli (hunk 11 above)
Comment (antalsz) [reply stdlib/iarrayLabels.mli:23]: We decided not to do this in the stdlib and only do this in Base

File: stdlib/iarrayLabels.mli (hunk 12 above)
Comment (antalsz) [reply stdlib/iarrayLabels.mli]: It turns out we haven't made enough of the `Array` API `local_` for these functions to be useful yet; deferred to the future.

File: stdlib/iarrayLabels.mli (hunk 18 above)
Comment (antalsz) [reply stdlib/iarrayLabels.mli:236]: These were the functions discussed above.

File: lambda/transl_array_comprehension.mli (hunk 29 above)
Comment (mshinwell) [reply lambda/transl_array_comprehension.mli:7]: I think this is correct.  All that should happen is that the middle-end won't be able to track anything about the contents of the array.

File: stdlib/iarray.ml (hunk 5 above)
Comment (mshinwell) [reply stdlib/iarray.ml:1]: Are these files completely new?  If so they should have JS copyright headers, you can crib from here: https://github.com/janestreet/base/blob/master/LICENSE.md

File: stdlib/iarray.ml [hunk 34]
```diff
@@ -0,0 +1,95 @@
...
+external ( .:() )     : 'a iarray -> int -> 'a = "%array_safe_get"
+external unsafe_get   : 'a iarray -> int -> 'a = "%array_unsafe_get"
+
+(* Immutable and mutable arrays have the same runtime representation; we
+   construct immutable arrays by constructing mutable arrays and then blindly
+   casting them to become immutable.  This is safe here because:
+   1. None of these functions mutate their array inputs;
+   2. None of these functions hold on to their array inputs; and
+   3. All of these functions return fresh arrays if they return an array. *)
+let init =
```
Comment (mshinwell) [inline stdlib/iarray.ml:23]: Personally I'd rather avoid uses of `Obj.magic` here -- there are very few uses in the stdlib, which seems like a good thing.  Can we not just copy the code of the original `Array` module and edit it as required (in which case the INRIA copyright notice must be preserved)?

File: stdlib/array.ml (hunk 1 above)
Comment (mshinwell) [reply stdlib/array.ml:165]: I would just write out the copying function, it's small, then we don't need `Obj.magic`.

File: lambda/translcore.ml (hunk 30 above)
Comment (mshinwell) [reply lambda/translcore.ml:35]: I'll read the changes to this module (hopefully tomorrow).
We should try to arrange things so that immutable arrays at the source level (except for ones constructed with comprehensions) turn into immutable arrays in Lambda -- this means that e.g. projection from array literals can be evaluated by the middle-end.  I'm unsure if this is happening at present.

Comment (mshinwell) [reply lambda/translcore.ml:35]: (and also it means that things like nested immutable array literals can be statically allocated)

File: lambda/lambda.mli [hunk 35]
```diff
@@ -38,15 +38,23 @@ type alloc_mode = private
   | Alloc_heap
   | Alloc_local
 
+type modify_mode = private
```
Comment (mshinwell) [inline lambda/lambda.mli:41]: I'm unsure why the change to add this `modify_mode` is in this PR.  Could we split this into a separate PR?

File: lambda/translcore.ml [hunk 36]
```diff
@@ -509,47 +503,58 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
         | cl ->
-            let imm_array =
+            let const =
               if Config.flambda2 then
-                Lprim (Pmakearray (kind, Immutable, mode), ll,
-                       of_location ~scopes e.exp_loc)
+                imm_array
               else
+                (* CR aspectorzabusky: Do we construct things correctly in this
+                   case? *)
```
Comment (mshinwell) [inline lambda/translcore.ml:557]: I think this is ok.

File: lambda/translcore.ml [hunk 37]
```diff
@@ -558,18 +563,21 @@ and transl_exp0 ~in_new_scope ~scopes e =
...
-              of_location ~scopes e.exp_loc)
+        makearray lambda_arr_mut
       end
   | Texp_list_comprehension comp ->
       let loc = of_location ~scopes e.exp_loc in
       Transl_list_comprehension.comprehension
         ~transl_exp ~scopes ~loc comp
-  | Texp_array_comprehension comp ->
+  | Texp_array_comprehension (_amut, comp) ->
+      (* We can ignore mutability here since we've already checked in in the
```
Comment (mshinwell) [inline lambda/translcore.ml:578]: Typo in comment

File: lambda/translcore.ml (hunk 30 above)
Comment (mshinwell) [reply lambda/translcore.ml:35]: @goldfirere @antalsz I've read these `Translcore` changes, I think they're ok - but we should test carefully, this stuff is tricky.

**Comment by antalsz** [comment]
This got resolved on ocaml-jst
